### ISSUE TITLE
Bulletproofs, Mirror Curves & Pairing Crypto

### ISSUE THREAD

**Issue by ChristopherA** [Chia-Network/bls-signatures#8]
I don't have broad knowledge of the capabilities & limitations of pairing crypto. Can zkp arithmetic circuits like Bulletproofs be constructed from this pairing curve? Is anyone working on support for arithmetic circuits using your choice of pairing curve?

Related, can this pairing curve support not just scalar integer Bulletproofs, but have a mirror curve with swap parameters to support signature Bulletproofs as per https://github.com/BlockchainCommons/secp256k1/issues/1 ?

**Comment by mariano54** [comment]
Yes, zcash is using this curve for zksnarks. Bulletproofs can be performed on this curve. It uses a 381 bit prime and a 255 bit group order for G1(which is a subgroup of the elliptic curve, so not all points on the curve are in the group).

According to Benedikt, a mirror curve is not necessary to perform EC math inside the bulletproof. All you need is a curve that has a prime = the order of the BLS12-381 group, but the order of the new curve does not have to be equal to the prime in BLS12-381. Then you could perform ec arithmetic in the circuit efficiently. However, it is not guaranteed that this new curve will be pairing friendly, so that is something to keep in mind.

**Comment by maaku** [comment]
The mirror curve is required if you want to support arbitrarily higher order bulletproofs (proofs of proofs of bulletproofs, etc.) without having to generate a new curve for each higher order.

**Comment by mariano54** [comment]
This library is focused on signatures, so I don't think we will be implementing any bulletproof features anytime soon.
